### ISSUE TITLE
Dependency Issue java.lang.NoSuchMethodError: com.monitorjbl.xlsx

### ISSUE THREAD

**Issue by Garagas93** [crealytics/spark-excel#135]
Hello, i'm trying to read a simple excel file but i keep getting an error. Here is the full stacktrace

`Exception in thread "main" java.lang.NoSuchMethodError: com.monitorjbl.xlsx.StreamingReader$Builder.open(Ljava/io/InputStream;)Lshadeio/poi/ss/usermodel/Workbook;
	at com.crealytics.spark.excel.StreamingWorkbookReader.openWorkbook(WorkbookReader.scala:56)
	at com.crealytics.spark.excel.WorkbookReader.withWorkbook(WorkbookReader.scala:14)
	at com.crealytics.spark.excel.WorkbookReader.withWorkbook$(WorkbookReader.scala:13)
	at com.crealytics.spark.excel.StreamingWorkbookReader.withWorkbook(WorkbookReader.scala:47)
	at com.crealytics.spark.excel.ExcelRelation.excerpt$lzycompute(ExcelRelation.scala:31)
	at com.crealytics.spark.excel.ExcelRelation.excerpt(ExcelRelation.scala:31)
	at com.crealytics.spark.excel.ExcelRelation.headerCells$lzycompute(ExcelRelation.scala:33)
	at com.crealytics.spark.excel.ExcelRelation.headerCells(ExcelRelation.scala:33)
	at com.crealytics.spark.excel.ExcelRelation.$anonfun$inferSchema$1(ExcelRelation.scala:148)
	at scala.Option.getOrElse(Option.scala:138)
	at com.crealytics.spark.excel.ExcelRelation.inferSchema(ExcelRelation.scala:147)
	at com.crealytics.spark.excel.ExcelRelation.<init>(ExcelRelation.scala:40)
	at com.crealytics.spark.excel.DefaultSource.createRelation(DefaultSource.scala:40)
	at com.crealytics.spark.excel.DefaultSource.createRelation(DefaultSource.scala:18)
	at com.crealytics.spark.excel.DefaultSource.createRelation(DefaultSource.scala:12)
	at org.apache.spark.sql.execution.datasources.DataSource.resolveRelation(DataSource.scala:318)
	at org.apache.spark.sql.DataFrameReader.loadV1Source(DataFrameReader.scala:223)
	at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:211)
	at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:178)
	at ExcelSparkParser$.readExcel(ExcelSparkParser.scala:86)
	at ExcelSparkParser$.$anonfun$main$3(ExcelSparkParser.scala:51)
	at ExcelSparkParser$.$anonfun$main$3$adapted(ExcelSparkParser.scala:48)
	at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:36)
	at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:33)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:198)
	at ExcelSparkParser$.main(ExcelSparkParser.scala:48)
	at ExcelSparkParser.main(ExcelSparkParser.scala)`

Envirorment:

- Scala 2.12.8 (tested also 2.12.0)
- Spark Core/SQL/Avro 2.4.0 (also tested on 2.4.3)
- Spark excel 0.12.0
- Running locally on my machine

I'm able to make it work only using an older version of spark-excel, to be precise version 0.11.2 , same spark versione and same scala version.  scala 2.11.x and spark-excel 0.11.1 are also working.
 
Thank you very much

**Comment by nightscape** [comment]
Did you explicitly specify
```scala
"com.monitorjbl" % "xlsx-streamer"
```
as a dependency?
If so, that would explain the problem.

**Comment by Garagas93** [comment]
Hi, no my sbt.build has only the following voices

`name := "ExcelParser"
version := "0.1"
scalaVersion := "2.12.8"

val sparkVersion = "2.4.0"

// https://mvnrepository.com/artifact/org.apache.spark/spark-core
libraryDependencies ++= Seq(
  "org.apache.spark" %% "spark-core" % sparkVersion ,
  "org.apache.spark" %% "spark-sql" % sparkVersion ,
  "org.apache.spark" %% "spark-avro" % sparkVersion 
)

// https://mvnrepository.com/artifact/com.crealytics/spark-excel
libraryDependencies += "com.crealytics" %% "spark-excel" % "0.11.2"`

**Comment by david-nguyen** [comment]
Received the same error as well using Azure Databricks, runtime: 5.5 (includes Apache Spark 2.4.3, Scala 2.11)

**Comment by wdorninger** [comment]
I get the same error when using the option . If you include the option maxRowsInMemory (mentioned in bold below) you get the above error. I'm new to spark-excel so maybe not entirely accurate but that was my observation. When removing that option the dependency error disappears.


import org.apache.spark.sql.SparkSession
import com.crealytics.spark.excel._

object ExcelApp {
  def main(args: Array[String]) {

    val spark = SparkSession.builder
      .appName("SparkExcel")
      .master("local[2]")
      .getOrCreate

    val df = spark.read.excel(
      useHeader = true,  
      **maxRowsInMemory = 20** 
    ).load("C:\\Temp\\excel-fun.xlsx")

    df.show()
  }
}

Exception in thread "main" java.lang.NoSuchMethodError: com.monitorjbl.xlsx.StreamingReader$Builder.open(Ljava/io/InputStream;)Lshadeio/poi/ss/usermodel/Workbook;

**Comment by nightscape** [comment]
@wdorninger which version are you using? If < 0.12.2, please try upgrading.

**Comment by wdorninger** [comment]
name := "spark-file"
version := "0.1"
scalaVersion := "2.11.12"

val sparkVersion = "2.1.2"

lazy val root = (project in file(".")).
  settings(
    name := "spark-file",
    version := "1.0",
    scalaVersion := "2.11.12",
    mainClass in Compile := Some("SimpleApp")
  )

libraryDependencies ++= Seq(
  "org.apache.spark" %% "spark-core" % sparkVersion,
  "org.apache.spark" %% "spark-sql" % sparkVersion,
  "org.apache.spark" %% "spark-mllib" % sparkVersion,
  "org.apache.spark" %% "spark-streaming" % sparkVersion,
  "org.apache.spark" %% "spark-hive" % sparkVersion,
  "com.crealytics" %% "spark-excel" % "0.12.3"
)

**Comment by david-nguyen** [comment]
still having the issue with 0.12.3 as well

**Comment by shestero** [comment]
Hi! I've got it.
Downgrade to 0.11.2 helps me.

name := "myname"
version := "0.1"
scalaVersion := "2.12.9"
val sparkVersion = "2.4.4"

libraryDependencies += "org.apache.spark" %% "spark-core" % sparkVersion
libraryDependencies += "org.apache.spark" %% "spark-sql" % sparkVersion

libraryDependencies += "com.crealytics" %% "spark-excel" % "0.11.2" // "0.12.3"

// MySQL
libraryDependencies += "mysql" % "mysql-connector-java" % "5.1.+"

retrieveManaged := true

**Comment by EnverOsmanov** [comment]
@nightscape here is reproducible [example](https://github.com/EnverOsmanov/sbt-assembly_TEST/tree/spark-excel) (branch name is `spark-excel`).

*sbt-missinglink* found too much missing links.
Queerly, current master (*0.12.4-SNAPSHOT*), published locally, runs the test project without any trouble. But *0.12.3* fails.

**Comment by nightscape** [comment]
@EnverOsmanov thanks for digging into this! How exactly did you publish locally?
There might be a difference between the fat jar (with shaded dependencies) that gets released to Maven Central and the slim jar (without dependencies) that gets published locally.

**Comment by EnverOsmanov** [comment]
In spark-excel project (**master** branch), I type `sbt` (for interactive mode) and then `++publishLocal` (**++** for Scala 2.11):
```
[info]  published spark-excel_2.11 to /home/enver/.ivy2/local/com.crealytics/spark-excel_2.11/0.12.4-SNAPSHOT/jars/spark-excel_2.11.jar
```
After that I use it in example project (**spark-excel** branch):
```
// libraryDependencies += "com.crealytics" %% "spark-excel" % "0.12.3"
libraryDependencies += "com.crealytics" %% "spark-excel" % "0.12.4-SNAPSHOT"
```
Intellij refreshs sbt dependencies and I run the `Hello.main` method with Intellij.

**Comment by carlo-quinonez** [comment]
When I use the `maxRowsInMemory` option, I see this error when reading a [simple excel file](https://github.com/ZuInnoTe/spark-hadoopoffice-ds/blob/master/src/it/resources/testsimple.xlsx?raw=true) on databricks 5.5 LTS (Spark 2.4.3, scala 2.11). I'm using the `com.crealytics:spark-excel_2.11:0.12.3` jar installed from Maven.

```python
excelPOC_df = spark.read.format("com.crealytics.spark.excel") \
        .option("useHeader", True) \
        .option("treatEmptyValuesAsNulls", True) \
        .option("inferSchema", True) \
        .option("sheetName", "Sheet1") \
        .option("maxRowsInMemory", 1) \
        .load("/mnt/scratch/excel/testsimple.xlsx")

display(excelPOC_df)
```

yields:
```
Py4JJavaError: An error occurred while calling o2138.load.
: java.lang.NoSuchMethodError: com.monitorjbl.xlsx.StreamingReader$Builder.open(Ljava/io/InputStream;)Lshadeio/poi/ss/usermodel/Workbook;
	at com.crealytics.spark.excel.StreamingWorkbookReader.openWorkbook(WorkbookReader.scala:56)
	at com.crealytics.spark.excel.WorkbookReader$class.withWorkbook(WorkbookReader.scala:14)
	at com.crealytics.spark.excel.StreamingWorkbookReader.withWorkbook(WorkbookReader.scala:47)
	at com.crealytics.spark.excel.ExcelRelation.excerpt$lzycompute(ExcelRelation.scala:31)
	at com.crealytics.spark.excel.ExcelRelation.excerpt(ExcelRelation.scala:31)
	at com.crealytics.spark.excel.ExcelRelation.headerCells$lzycompute(ExcelRelation.scala:33)
	at com.crealytics.spark.excel.ExcelRelation.headerCells(ExcelRelation.scala:33)
	at com.crealytics.spark.excel.ExcelRelation$$anonfun$7.apply(ExcelRelation.scala:108)
	at com.crealytics.spark.excel.ExcelRelation$$anonfun$7.apply(ExcelRelation.scala:106)
	at scala.Option.getOrElse(Option.scala:121)
	at com.crealytics.spark.excel.ExcelRelation.headerColumns$lzycompute(ExcelRelation.scala:106)
	at com.crealytics.spark.excel.ExcelRelation.headerColumns(ExcelRelation.scala:105)
	at com.crealytics.spark.excel.ExcelRelation$$anonfun$inferSchema$1.apply(ExcelRelation.scala:164)
	at com.crealytics.spark.excel.ExcelRelation$$anonfun$inferSchema$1.apply(ExcelRelation.scala:162)
	at scala.Option.getOrElse(Option.scala:121)
	at com.crealytics.spark.excel.ExcelRelation.inferSchema(ExcelRelation.scala:162)
	at com.crealytics.spark.excel.ExcelRelation.<init>(ExcelRelation.scala:36)
	at com.crealytics.spark.excel.DefaultSource.createRelation(DefaultSource.scala:40)
	at com.crealytics.spark.excel.DefaultSource.createRelation(DefaultSource.scala:18)
	at com.crealytics.spark.excel.DefaultSource.createRelation(DefaultSource.scala:12)
	at org.apache.spark.sql.execution.datasources.DataSource.resolveRelation(DataSource.scala:343)
	at org.apache.spark.sql.DataFrameReader.loadV1Source(DataFrameReader.scala:307)
	at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:293)
	at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:214)
	at sun.reflect.GeneratedMethodAccessor435.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:380)
	at py4j.Gateway.invoke(Gateway.java:295)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:251)
	at java.lang.Thread.run(Thread.java:748)
```

Without the `maxRowsInMemory` option, it's successful:

```
excelPOC_df = spark.read.format("com.crealytics.spark.excel") \
        .option("useHeader", True) \
        .option("treatEmptyValuesAsNulls", True) \
        .option("inferSchema", True) \
        .option("sheetName", "Sheet1") \
        .load("/mnt/scratch/excel/testsimple.xlsx")

display(excelPOC_df)
```

yields:
![Screen Shot 2019-12-02 at 5 22 47 PM](https://user-images.githubusercontent.com/53790047/70012140-5d839100-1528-11ea-92d2-94bdd3421f0a.png)

**Comment by xxxatt** [comment]
same issue here when enable the maxRowsInMemory option

**Comment by b003k** [comment]
same issue with maxRowsInMemory. 
Need this option due to the excel file being too big and throwing "zip bomb detected" in databricks.

**Comment by pmario26** [comment]
I am facing the same issue when the maxRowsInMemory option is enabled.
My scala version is 2.11.8, i tried using the spark-excel 0.12.2 and 0.12.3 and had the same error.

**Comment by EnverOsmanov** [comment]
Guys, don't hesitate to test the snapshot below. I hope your results will encourage spark-excel's maintainer to release a fix sooner.
```
resolvers +=
  "Sonatype OSS Snapshots" at "https://oss.sonatype.org/content/repositories/snapshots"

libraryDependencies += "com.github.enverosmanov" %% "spark-excel" % "0.12.4-SNAPSHOT"
```

**Comment by nightscape** [comment]
@EnverOsmanov thanks for the work! Everybody please test this, if it works for other people I'd be happy to merge. Just haven't had the time to test it myself.

**Comment by carlo-quinonez** [comment]
I installed the JAR into Databricks by using the coordinate `com.github.enverosmanov:spark-excel_2.11:0.12.4-SNAPSHOT` and using `https://oss.sonatype.org/content/repositories/snapshots/
` as the repository URL.

It still fails, but the error is different. Now I get a simple `The spark driver has stopped unexpectedly and is restarting. Your notebook will be automatically reattached.`.  I'll retry on a bigger cluster with more memory to see if that helps.

**Comment by carlo-quinonez** [comment]
So `com.github.enverosmanov:spark-excel_2.11:0.12.4-SNAPSHOT` imports my excel file successfully, but I don't see any data in the resulting dataframe. There's 200k+ rows in the excel file, but the preview in Databricks only shows 1 row filled with all `null`. Oddly, the `.count()` on the dataframe returns correct number of rows, but counting all the rows took as long as the initial import - is that expected?

![Screen Shot 2019-12-20 at 10 18 30 AM](https://user-images.githubusercontent.com/53790047/71281203-24ae2f00-2312-11ea-8f58-fadf3158cf9b.png)

Using the `maxRowsInMemory` options works fine with the tiny test excel file I have. The preview of the dataframe for this file DOES show all the expected data - even when I set `maxRowsInMemory` to `2`.

![Screen Shot 2019-12-20 at 10 19 51 AM](https://user-images.githubusercontent.com/53790047/71281240-47404800-2312-11ea-9d95-2781d5a306f8.png)

**Comment by EnverOsmanov** [comment]
Thank you, @carlo-quinonez 
I had an empty result dataset too on one of my test examples, but all of others worked perfectly. And it was fixed by specifying schema explicitly. This problem only happened with **maxRowsInMemory** enabled.
```
import org.apache.spark.sql.Encoders

case class Foo(A: Int, B: Int, C: Int, D: Int) 

val dataframe = spark.read                                
  .excel(workbookPassword = "fooba", maxRowsInMemory = 20)
  .schema(Encoders.product[Foo].schema)                   
  .load(path)                                             
                                                          
dataframe.show(truncate = false)                          
```

**Comment by nightscape** [comment]
I've released `0.12.4` with `xlsx-streamer` shaded. This should hopefully fix this issue.
If it doesn't, please reopen.
